### DISCUSSION TITLE
Implementing Frequency Control for Nidec BLDC Fan Motor in Tasmota with ESP32

### DISCUSSION THREAD

**Discussion by vanhpham** [arendst/Tasmota#21911] [Q&A]
Hello,

I'm trying to control my fan with Tasmota, but I'm facing an issue. My fan, which is a Nidec BLDC motor, is controlled by a clock signal (CLK) with a frequency range from 150Hz to 380Hz. However, in Tasmota's GPIO configuration, I only see the option for PWM, which doesn't work correctly for my fan. When the duty cycle exceeds 30%, the fan speed maxes out.

Here is the previous code that works fine with my fan:
```
const int ledChannel = 0;
int freq = 0;
const int resolution = 8;

void setup() {
  ledcSetup(ledChannel, freq, resolution);
  ledcAttachPin(CLKpin, ledChannel);
  ledcWriteTone(ledChannel, freq);
}
```
I want to modify the source code of Tasmota to implement my own function to control the fan using the CLK signal. However, I'm not sure how to proceed with this modification.

Thank you!

**Comment by sfromis** [comment]
You can control the duty cycle of a Tasmota PWM output, using the `Dimmer` command if you leave it as a "light" configuration, or the `PWM` command if in basic mode. And you have the `PWMFrequency` command to select your frequency.

With ESP32 and Berry, you also have a simple API for PWM duty cycle, phase and frequency.

**Reply by vanhpham** [reply]
I'm new to tasmota and have no idea about this, i read some instruction how to modify src code and create a custom binary file but still dont get clear. I also try berry but doc about this is very limitation or maybe im not searching right. Can you give me some clear instruction of make a Berry script that run with my setup. Thank u 🥺

**Reply by sfromis** [reply]
What I posted is that you do not need to modify source code to control the duty cycle, if you take advantage of the existing Tasmota features, either for PWM or with Berry.

The Berry API I mentioned is on https://tasmota.github.io/docs/Berry/#gpio-module but as with any language, there is a learning curve to get started, and it also depends a lot on how you envision controlling the fan. Maybe a Tasmota command (created with Berry) could be a way, but that depends on what you want to integrate with.
